api/admin/controller/collection_settings.py

```diff
@@ -0,0 +1,266 @@
+from nose.tools import set_trace
+from . import SettingsController
+import flask
+from flask import Response
+from flask_babel import lazy_gettext as _
+import json
+from core.opds_import import (OPDSImporter, OPDSImportMonitor)
+from api.admin.problem_details import *
+from api.feedbooks import FeedbooksOPDSImporter
+from api.opds_for_distributors import OPDSForDistributorsAPI
+from api.overdrive import OverdriveAPI
+from api.odilo import OdiloAPI
+from api.bibliotheca import BibliothecaAPI
+from api.axis import Axis360API
+from api.rbdigital import RBDigitalAPI
+from api.enki import EnkiAPI
+from api.odl import ODLWithConsolidatedCopiesAPI, SharedODLAPI
+from core.model import (
+    Collection,
+    ConfigurationSetting,
+    ExternalIntegration,
+    get_one,
+    get_one_or_create,
+    Library,
+)
+from core.util.problem_detail import ProblemDetail
+
+class CollectionSettingsController(SettingsController):
+
+    # SET UP
+    def process_collections(self):
+        self.set_up_protocols()
+        if flask.request.method == 'GET':
+            return self.process_get()
+        else:
+            return self.process_post()
+
+    def set_up_protocols(self):
+        self.protocols = self._get_collection_protocols(self.PROVIDER_APIS)
+
+        # If there are storage integrations, add a mirror integration
+        # setting to every protocol's 'settings' block.
+        mirror_integration_setting = self._mirror_integration_setting()
+        if mirror_integration_setting:
+            for protocol in self.protocols:
+                protocol['settings'].append(mirror_integration_setting)
```

Comment:
This is a little trickier than other protocol setup steps. We could start off with a site that has no mirror integration, and then the admin creates one. However, if the admin creates a new integration, the entire site configuration is reloaded, which means a new `CollectionSettingsController` will be created. Move this whole method into the `CollectionSettingsController` constructor and we'll be fine.

Reply:
We tried this out and it broke a bunch of tests, because the test setup creates the CirculationManager with no mirror integration in place. We decided to avoid setting self.protocols for this one -- we can always come back to this later.